beacon_chain/state/state_transition.py

```diff
@@ -330,26 +348,64 @@ def fill_recent_block_hashes(active_state: ActiveState,
     )
 
 
-def compute_cycle_transitions(
-        crystallized_state: CrystallizedState,
-        active_state: ActiveState,
-        block: 'Block',
-        config: Dict[str, Any]=DEFAULT_CONFIG) -> Tuple[CrystallizedState, ActiveState]:
-    while block.slot_number >= crystallized_state.last_state_recalc + config['cycle_length']:
-        crystallized_state, active_state = initialize_new_cycle(
-            crystallized_state,
-            active_state,
-            block,
-            config=config,
-        )
-        if ready_for_dynasty_transition(crystallized_state, block, config):
-            crystallized_state = compute_dynasty_transition(
-                crystallized_state,
-                block,
-                config
-            )
+def apply_rewards_and_penalties(crystallized_state: CrystallizedState,
+                                active_state: ActiveState,
+                                block: 'Block',
+                                config: Dict[str, Any]=DEFAULT_CONFIG) -> List['ValidatorRecord']:
+    validators = crystallized_state.validators
+    updated_validators = deepcopy(crystallized_state.validators)
+    active_validator_indices = get_active_validator_indices(
+        crystallized_state.current_dynasty,
+        validators
+    )
 
-    return crystallized_state, active_state
+    time_since_finality = block.slot_number - crystallized_state.last_finalized_slot
+    total_deposits = crystallized_state.total_deposits
+    reward_quotient = config['base_reward_quotient'] * int(sqrt(total_deposits))
+    quadratic_penalty_quotient = int(sqrt(config['sqrt_e_drop_time'] / config['slot_duration']))
+
+    last_state_recalc = crystallized_state.last_state_recalc
+    block_vote_cache = active_state.block_vote_cache
+
+    # FFG Rewards
+    for slot in range(last_state_recalc - config['cycle_length'], last_state_recalc):
+        block = active_state.chain.get_block_by_slot_number(slot)
+        if block:
+            block_hash = block.hash
+            total_participated_deposits = block_vote_cache[block_hash]['total_voter_deposits']
+            voter_indices = block_vote_cache[block_hash]['total_voter_deposits']
+        else:
+            total_participated_deposits = 0.0
+            voter_indices = set()
+
+        if time_since_finality <= 2 * config['cycle_length']:
+            for index in active_validator_indices:
+                if index in voter_indices:
+                    updated_validators[index].balance = int(
+                        validators[index].balance +
+                        validators[index].balance *
+                        (1 / reward_quotient) *
+                        (2 * total_participated_deposits / total_deposits - 1)
```

Comment:
It doesn't follow the latest spec.

Before Sep, 18
* Participating validators gain `B * reward_factor * (2 * total_participated_deposits / total_deposits - 1)`

Now:
* Participating validators gain `B // reward_quotient * (2 * total_participated_deposits - total_deposits) // total_deposits` (note: this may be negative)

Reply:
In the new formula, `reward_quotient` is an integer, it seems the result should be an integer too, so no need to use `int()`?

Reply:
yeah, I forgot to fix the logic. Should be good now